OSLObject FaceVarying and Uniform interpolation

donboie opened the pull request:
This PR allows OSLObject to shade using _Uniform_ and _FaceVarying_ interpolations in addition to the preexisting  _Vertex_  mode.

I'm concerned with the performance regression when only reading already vertex interpolated variables. I have another branch were I query the variables read in the shader but I'm unsure at what point I should read these and set the names plug on the internal resample node. Hopefully performance hit is acceptable and we can address the performance in a subsequent PR?

johnhaddon commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -70,7 +71,25 @@
+
+			"description",
+			"""
+			interpolation mode in which to process the object. 
```
Capitalise, remove blank lines.

donboie commented on python/GafferOSLTest/OSLObjectTest.py:
```diff
@@ -236,5 +236,137 @@ def testReferencePromotedPlug( self ) :
+		self.assertEqual( cubeObject["P_copy"].data[4], IECore.V3f(  0.0,  0.5,  0.0 ) + IECore.V3f( 1, 2, 3 ))
+		self.assertEqual( cubeObject["P_copy"].data[5], IECore.V3f(  0.0, -0.5,  0.0 ) + IECore.V3f( 1, 2, 3 ))
+
+def testCanShadeFaceVaryingInterpolatedPrimitiveVariablesAsVertex( self ) :
```
I should have squashed that intermediate step - sorry. I'm a little concerned sometimes I'm too eager to squash changes into fewer commits  but in this case it would have made things clearer.

I'm unclear if I need to update affects & hash to consider the resampledInPlug?

johnhaddon replied:
> I'm unclear if I need to update affects & hash to consider the resampledInPlug?

The best rule of thumb is to work backwards from the `compute()`, since that's the thing you're actually interested in implementing, and the rest is just an annoyance needed to keep caching and dirty propagation working. You only need to implement `compute()` for output plugs, and only then if they don't have an input connection. Since `resampledInPlug()` is an input plug _and_ has an input connection, you definitely don't need to implement compute for it and therefore don't need to compute a `hash()` for it, or add it to the affected outputs in `affects()`.

What I was trying to get it is that because you now access a value from `resampledInPlug()` in `computeProcessedObject()`, you also need to include it in the hash in `hashProcessedObject()` and ensure that it causes `out.object` to be dirtied in `affects()`. You're probably getting away with it for now because `interpolationPlug()` is approximating the same effect, but as we make the logic for which variables are resampled more complex, getting this right will become more important.

donboie replied:
Thanks for the explanation

donboie commented:
Hopefully I've addressed all of your comments apart from the Constant interpolation. I attempted to support this and noticed we'd either have convert the ScalarData primvars to VectorData in OSLObject for use by the shading engine or update the shading engine to treat scalars as single element vectors.

I am struggling a little to think of examples where we'd want to shade constant resampled primvars so how do you feel about deferring this until we have a clear need of it?

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -169,6 +225,7 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
 
 	interpolationPlug()->hash( h );
 	h.append( inPlug()->fullTransformHash( path ) );
+	h.append( resampledInPlug()->fullAttributesHash( path ) );
```
I think we've got our wires crossed somewhere - in `computeProcessedObject()` we call `resampledInPlug()->objectPlug()->getValue()`, so here we need to call `resampledInPlug()->objectPlug()->hash( h )`, and _not_ anything to do with the attributes.

Recap :

- A ScenePlug has various child plugs each representing a property of a location
    - `objectPlug()` represents geometry/cameras etc
    - `transformPlug()` represents the local transform
    - `boundPlug()` represents the local bounding box
    - `attributesPlug()` represents the local attributes. Note that "attributes" here is used in the RenderMan sense of inheritable properties that are orthogonal to geometry/transforms etc. This is different to the "everything is an attribute" approach taken by certain other packages.
- The `fullAttributes/fullAttributesHash` methods walk all parent locations, getting the local attributes (or their hash) and combining them with the result so far. They're rarely needed in a computation context.

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -218,9 +264,50 @@ IECore::ConstObjectPtr OSLObject::computeProcessedObject( const ScenePath &path,
+
+		if (!prim)
+		{
+			static_cast<StringPlug *>( output )->setValue( "" );
```
`output->setToDefault()` might be _slightly_ better here, because it can reuse the internal default value rather than needing to construct a new one.

johnhaddon commented on python/GafferOSLUI/OSLObjectUI.py:
```diff
@@ -70,7 +71,23 @@
+			"description",
+			"""
+			Interpolation mode in which to process the object. 
+			All primitive variables are resampled to match the selected interpolation. 
```
All non-constant primitive variables. Maybe we could explain a bit better what we mean by "in which to process the object"  as well? How about something like this?

"""
The interpolation type of the primitive variables created by this node. For instance, Uniform interpolation means that the shader is run once per face on a mesh, allowing it to output primitive variables with a value per face. All non-constant input primitive variables are resampled to match the selected interpolation so that they can be accessed from the shader.
"""

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -151,6 +166,8 @@ void OSLObject::hashProcessedObject( const ScenePath &path, const Gaffer::Contex
 		shader->attributesHash( h );
 	}
+
+	interpolationPlug()->hash( h );
```
Thanks for squashing that middle commit down. One note for the future - getting affects/hash/compute to match is one of the most important things in coding a node, so it's best not to split the hash/compute changes across commits like this as it makes review hard. I'd have been quite happy to see this commit squashed into the next too...

donboie commented:
I really *hope* I understand now :man_facepalming: .Thanks so much for detailed explanations and patience. 

Are you happy with leaving the constant interpolation for the future?

johnhaddon commented on src/GafferOSL/OSLObject.cpp:
```diff
@@ -88,21 +120,43 @@ const Gaffer::IntPlug *OSLObject::interpolationPlug() const
-	}
-	else if( input == inPlug()->transformPlug() )
-	{
+		outputs.push_back( outPlug()->objectPlug() );
```
There seem to be three calls to `outputs.push_back( outPlug()->objectPlug() )` here - we just need the one, as in my example from the previous review.
